pkg/yarn/copilot/nm/nm.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/pleg"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/util/system"
+	"github.com/koordinator-sh/koordinator/pkg/yarn/copilot/utils"
+)
+
+type NodeMangerOperator struct {
+	CgroupRoot string
+	CgroupPath string
+
+	SyncMemoryCgroup bool
+
+	containerWatch pleg.Watcher
+	NMEndpoint     string //localhost:8042
+	client         *resty.Client
+	ticker         *time.Ticker
+}
+
+func NewNodeMangerOperator(cgroupRoot string, cgroupPath string, syncMemoryCgroup bool, endpoint string, syncPeriod time.Duration) (*NodeMangerOperator, error) {
+	watcher, err := pleg.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	cli := resty.New()
+	cli.HostURL = endpoint
+	return &NodeMangerOperator{
+		CgroupRoot:       cgroupRoot,
+		CgroupPath:       cgroupPath,
+		SyncMemoryCgroup: syncMemoryCgroup,
+		containerWatch:   watcher,
+		NMEndpoint:       endpoint,
+		client:           cli,
+		ticker:           time.NewTicker(syncPeriod),
+	}, nil
+}
+
+func (n *NodeMangerOperator) Run(stop <-chan struct{}) error {
+	if n.SyncMemoryCgroup {
+		return n.syncMemoryCgroup(stop)
+	}
+	return nil
+}
+
+func (n *NodeMangerOperator) syncMemoryCgroup(stop <-chan struct{}) error {
+	if err := n.containerWatch.AddWatch(filepath.Join(n.CgroupRoot, system.CgroupCPUDir, n.CgroupPath)); err != nil {
+		return err
+	}
+	if err := n.ensureMemoryCgroupDir(); err != nil {
+		return err
+	}
+	for {
+		select {
+		case event := <-n.containerWatch.Event():
+			switch pleg.TypeOf(event) {
+			case pleg.DirCreated:
+				n.createMemoryCgroup(event.Name)
+			case pleg.DirRemoved:
+				n.removeMemoryCgroup(event.Name)
+			default:
+				klog.V(5).Infof("skip %v unknown event", event.Name)
+			}
+		case <-n.ticker.C:
+			n.syncAllCgroup()
+		case <-stop:
+			return nil
+		}
+	}
+}
+
+func (n *NodeMangerOperator) syncAllCgroup() {
+	subDirFunc := func(dir string) map[string]struct{} {
+		res := map[string]struct{}{}
+		filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				klog.Warningf("ignore file %s error:%s", path, err.Error())
+				return err
+			}
+			if info.IsDir() && path != dir {
+				res[path] = struct{}{}
+				return filepath.SkipDir
+			}
+			return nil
+		})
+		return res
+	}
+	cpuList := subDirFunc(filepath.Join(n.CgroupRoot, system.CgroupCPUDir, n.CgroupPath))
+	memList := subDirFunc(filepath.Join(n.CgroupRoot, system.CgroupMemDir, n.CgroupPath))
+	toCreate, toDelete := utils.DiffMap(cpuList, memList)
+	for path, _ := range toCreate {
+		n.createMemoryCgroup(path)
+	}
+	for path, _ := range toDelete {
+		n.removeMemoryCgroup(path)
+	}
+}
+
+func (n *NodeMangerOperator) removeMemoryCgroup(fileName string) {
+	basename := filepath.Base(fileName)
+	if !strings.HasPrefix(basename, "container_") {
+		klog.V(5).Infof("skip file %s, which is not a yarn container file", basename)
+		return
+	}
+	memCgroupPath := filepath.Join(n.CgroupRoot, system.CgroupMemDir, n.CgroupPath, basename)
```

Comment:
there are some utils in koordlet can be used, which can be also supports cgroupV2